### ISSUE TITLE
False SC2269 (info): This variable is assigned to itself

### ISSUE THREAD

**Issue by yaitskov** [koalaman/shellcheck#3411]
#### For bugs with existing features

- Rule Id SC2269
- My shellcheck version 0.11.0

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
while read -r _default _via GWIP _dev GWNIC _rest ; do                                                                             
      GWIP="$GWIP"                                                                                                                       
      GWNIC="$GWNIC"                                                                                                                     
      echo "GWIP = $GWIP; GWNIC = $GWNIC"                                                                                                  
done < <(ip route list | grep '^default') 
```

#### Here's what shellcheck currently says:

```
In /nix/store/hvzlcx8j2y98qnk1ga0kr30w8i5cbsfp-vpn-routing/bin/vpn-routing line 18:
  GWIP="$GWIP"
  ^--^ SC2269 (info): This variable is assigned to itself, so the assignment does nothing.


In /nix/store/hvzlcx8j2y98qnk1ga0kr30w8i5cbsfp-vpn-routing/bin/vpn-routing line 19:
  GWNIC="$GWNIC"
  ^---^ SC2269 (info): This variable is assigned to itself, so the assignment does nothing.
```

#### Here's what I wanted or expected to see:

There is no issue should be raised.
Without such assignment variable cannot leak the scope of the while loop body.

**Comment by cD1rtX3** [comment]
What do you mean by "[w]ithout such assignment variable cannot leak the scope of the while loop body"? I don't think that's the case.

**Comment by yaitskov** [comment]
let me rephrase: read introduces local variables and the statement that checker flags as meaningless is just extending variable scope. 

Is there a keyword for tuning variable scope? 

e.g. nix has inherit

**Comment by cD1rtX3** [comment]
Aren't variables always unscoped, excepting `export`s? I'm not familiar with non-POSIX sh, so I'm not sure about nix.

**Comment by yaitskov** [comment]
I use regular bash and variables introduced via read keyword are scoped, as practice shows.

**Comment by cD1rtX3** [comment]
??

<img width="373" height="118" alt="Image" src="https://github.com/user-attachments/assets/5b644a03-3a7f-48c3-a0fb-5f493209c5a2" />

**Comment by yaitskov** [comment]
![IMG_20260324_204737236_HDR~2.jpg](https://github.com/user-attachments/assets/8a901672-6a3c-4dc0-83c1-6d71a1f20ce0)

**Comment by e-kwsm** [comment]
You are using pipeline.
Variables are *scoped* unless lastpipe is enabled.

**Comment by cD1rtX3** [comment]
Right. You're piping into a command, which starts a subshell.

For example:

<img width="268" height="35" alt="Image" src="https://github.com/user-attachments/assets/2a09b113-11d8-4441-8db6-a80e6d037971" />

The redundant assignment doesn't actually do anything, since it's in a subshell:

<img width="490" height="75" alt="Image" src="https://github.com/user-attachments/assets/c64bd955-35a9-47b0-b8eb-94b525e3a281" />

**Comment by cD1rtX3** [comment]
Also, don't point a camera at your screen to take a screenshot.

**Comment by wileyhy** [comment]
Okay. For reference: dynamic scoping of variables (global vs. local), and
use of variables within subshells, ie, `(...)`, `$(...)`, pipelines,
background jobs `&`, etc., are two separate things.

Local variables in bash are created with the shell built-in `local`. dash
and BusyBox sh, as well as zsh, yash and oksh also have a `local` built-in.
ksh does not have `local`, nor does POSIX-2024 specify a `local` built-in.
(I just tested these statements at a command line and looked at the Posix
spec.).

If a shell has the `local` built-in, then its subshells can have both
global and local variables.

In bash and in Linux, a subshell is a command execution environment in a
parent/child relationship with another command execution environment. A
child shell, ie, a subshell, inherits (I believe) an exact copy of the
parent shell's environment, but after a subshell has been created one shell
cannot alter the environment of another shell. That's how subshells can
seem to have dynamic scope.

The `local` built-in allows dynamic scope within just one execution
environment. In bash, the `local` built-in can only be used in functions,
and is limited to within the function where it's defined. "$Foo" can have a
global value of "bar", then within a function "$Foo" can have a local value
of "crazy", then after the function returns "$Foo"s previous value is
visible again.

If you unset a local variable (again, in bash) within a function, the
global variable will remain masked until the function returns. oksh and I
don't know which others makes the global scope visible when a local
variable is unset.
